### PR TITLE
Update quadrature for improved safety & robustness

### PR DESCRIPTION
**Pull request by geogunow**
This PR restructures the `Quadrature` class in OpenMOC so that it is:
- More extensible
- Easier to comprehend
- Safer with respect to memory allocations and management

In its current form, the `Quadrature` class uses quite a bit of C-style low level memory management. This is problematic because it can lead to memory leaks or segmentation faults with just small changes to the code. This PR changes the `Quadrature` class to start using more high-level C++ data structures. In general, high-level C++ data structures provide robust and efficient memory access patterns if the context can be sufficiently expressed with the available semantics. This is certainly true of the `Quadrature` class where the only performance-critical regions are the `getWeightInline` and `getSinThetaInline` functions.

For this PR, moving to higher-level C++ data structures largely means changing from explicitly allocated array to C++ `std::vector`. This allows for destructors to be removed, as C++ vectors automatically manage their memory. In addition to reducing the code complexity, this improves robustness in the face of exceptions. In the current code, exceptions could cause memory leaks since the members are not self-aware of their memory consumption (ie only `Quadrature` is aware), causing potential memory leaks or segmentation faults in the destructor if exceptions are thrown before the end of initialization.

This PR also cleans up some signed-unsigned comparisons, simplifies printing, and reduces the amount of magic numbers (though further work is needed).

### CODE REVIEW

File: src/Quadrature.cpp [hunk 1]
```diff
@@ -1362,89 +1236,89 @@ std::vector <double> GLPolarQuad::getLegendreRoots(int n) {
 
 
 /**
- * @brief Calculates the weights to be used in Gauss-Legendre Quadrature.
+ * @brief calculates the weights to be used in Gauss-Legendre Quadrature
```
Comment (GiudGiud) [inline src/Quadrature.cpp]: Keep previous form

File: src/Quadrature.cpp [hunk 2]
```diff
@@ -1362,89 +1236,89 @@ std::vector <double> GLPolarQuad::getLegendreRoots(int n) {
...
-       * legendrePolynomial(n-1, roots[i]));
+        (n*n*legendrePolynomial(n-1, roots[i])
+         *legendrePolynomial(n-1, roots[i]));
     weights.push_back(value);
   }
 
   return weights;
 }
 
-
```
Comment (GiudGiud) [inline src/Quadrature.cpp]: Keep spacing

File: src/Quadrature.cpp [hunk 3]
```diff
@@ -1362,89 +1236,89 @@ std::vector <double> GLPolarQuad::getLegendreRoots(int n) {
...
-  int n = nodes.size();
+  size_t n = nodes.size();
 
   std::vector<double> weights;
 
-  // declare an array to store the elements of the augmented-matrix
-  long double A[n][n+1];
+  // declare a vector to store the elements of the augmented-matrix
+  std::vector< std::vector<long double> > A(n);
+  for (size_t i=0; i < n; ++i) {
```
Comment (GiudGiud) [inline src/Quadrature.cpp]: use resize2D

File: src/Quadrature.cpp [hunk 4]
```diff
@@ -1459,40 +1333,38 @@ std::vector<double> GLPolarQuad::getCorrectedWeights(int azim) {
...
 
   // fill the vector of weights so that it is indexed to the original
   // vector of nodes
-  for (int i=0; i<n; ++i)
+  for (size_t i=0; i<n; ++i)
     weights[index[i]] = double(x[i]);
 
-  return weights;
-
+   return weights;
```
Comment (GiudGiud) [inline src/Quadrature.cpp]: delete extra white space

**Comment by GiudGiud** [comment]
@geogunow Resolved the merge conflicts that my PR created. Hope this helps.

File: src/Quadrature.h [hunk 5]
```diff
@@ -29,17 +29,44 @@
...
   GAUSS_LEGENDRE,
   EQUAL_WEIGHT,
   EQUAL_ANGLE
 };
 
+/** Shorthand for vectors of floats */
+typedef std::vector<double> DoubleVec;
+typedef DoubleVec::const_iterator DVCI;
+
+template <typename T>
```
[the remark this answers is not in the document]
Comment (geogunow) [reply src/Quadrature.h]: Not easy to overwrite `printf`. But overwriting the `<<` operator is easier. Notice that there is still a `toString` method.

File: src/Quadrature.h [hunk 6]
```diff
@@ -259,21 +278,37 @@ class EqualAnglePolarQuad: public Quadrature {
...
+
+
 /**
  * @brief Returns the total weight for Tracks with the given azimuthal and
  *        polar indexes without error checking and inlined
  * @details Angular weights are multiplied by Track spacings
- * @param azim index of the azimuthal angle of interest
- * @param polar index of the polar angle of interest
+ * @param azim index of the azimuthal angle of size_terest
+ * @param polar index of the polar angle of size_terest
```
[the remark this answers is not in the document]
Comment (geogunow) [reply src/Quadrature.h]: oops, find and replace got me

**Comment by GiudGiud** [comment]
All good ! Merging

**Pull request merged**